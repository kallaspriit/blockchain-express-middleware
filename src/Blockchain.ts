import Axios from "axios";
import * as querystring from "querystring";
import { dummyLogger, ILogger } from "./abstract-logger";
import { Invoice } from "./index";

/**
 * Base configuration that has reasonable defaults and do no require to be redefined by the user.
 */
export interface IBlockchainBaseConfig {
  apiBaseUrl: string;
  gapLimit?: number;
}

/**
 * Configuration required to be provided by the user.
 */
export interface IBlockchainRequiredConfig {
  apiKey: string;
  xPub: string;
}

/**
 * Configuration that the user provides, includes mandatory user configuration and optional base configuration.
 */
export type IBlockchainUserConfig = Partial<IBlockchainBaseConfig> & IBlockchainRequiredConfig;

/**
 * Combined configuration including base and user provided configurations.
 */
export type IBlockchainConfig = IBlockchainBaseConfig & IBlockchainRequiredConfig;

/**
 * Parameters for the generate receiving address endpoint.
 */
export interface IGenerateReceivingAddressParameters {
  xpub: string;
  callback: string;
  key: string;
  gap_limit?: number;
}

/**
 * Response for the generate receiving address endpoint.
 */
export interface IReceivingAddress {
  address: string;
  index: number;
  callback: string;
}

/**
 * Parameters for creating an invoice.
 */
export interface ICreateInvoiceInfo {
  dueAmount: number;
  message: string;
  secret: string;
  callbackUrl: string;
}

/**
 * Default base configuration.
 */
export const defaultBaseConfig: IBlockchainBaseConfig = {
  apiBaseUrl: "https://api.blockchain.info/v2/receive",
};

/**
 * Provides API for receiving payments through blockchain.info service.
 *
 * To get the extended public key (xPub), copy your mnemonic words to http://bip32.org/ "Passphrase" field, wait for it
 * to generate the extended key and use the value of "Derived Public Key" as the "xPub" parameter.
 *
 * See https://blockchain.info/api/api_receive for API documentation.
 */
export default class Api {
  private readonly config: IBlockchainConfig;

  /**
   * Constructor.
   *
   * Accepts configuration and optional logger to use.
   *
   * @param userConfig User configuration (can override base configuration as well)
   * @param log Logger to use (defaults to console, but you can use bunyan etc)
   */
  public constructor(userConfig: IBlockchainUserConfig, private readonly log: ILogger = dummyLogger) {
    this.config = {
      ...defaultBaseConfig,
      ...userConfig,
    };
  }

  /**
   * Generates a new receiving address.
   *
   * @param callbackUrl URL to call on new transactions and confirmation count changes
   */
  public async generateReceivingAddress(callbackUrl: string): Promise<IReceivingAddress> {
    const { apiBaseUrl, xPub, apiKey, gapLimit } = this.config;
    const parameters: IGenerateReceivingAddressParameters = {
      xpub: xPub,
      callback: callbackUrl,
      key: apiKey,
    };

    // only add the gap limit parameters if it has been set
    if (gapLimit !== undefined) {
      parameters.gap_limit = gapLimit;
    }

    // build the request url
    const url = `${apiBaseUrl}?${querystring.stringify(parameters)}`;

    this.log.info(
      {
        parameters,
        url,
      },
      "generating receiving address",
    );

    // attempt to generate the receiving address (throws for non 2xx response)
    try {
      const response = await Axios.get<IReceivingAddress>(url);

      return response.data;
    } catch (error) {
      // log failure and rethrow the error
      this.log.error(
        {
          message: error.message,
          status: error.response.status,
          statusText: error.response.statusText,
          data: error.response.data,
          parameters,
          url,
        },
        "generating receiving address failed",
      );

      throw error;
    }
  }

  /**
   * Creates a new invoice.
   *
   * First generates the receiving address and then the invoice.
   *
   * @param info Invoice info
   */
  public async createInvoice({ dueAmount, message, secret, callbackUrl }: ICreateInvoiceInfo): Promise<Invoice> {
    // build invoice signature to later verify that the handle payment request is valid
    const signature = Invoice.getInvoiceSignature(
      {
        dueAmount,
        message,
      },
      secret,
    );

    // build the callback url, containing invoice info signed with the secret
    const decoratedCallbackUrl = `${callbackUrl}?${querystring.stringify({ signature })}`;

    // generate next receiving address
    const receivingAddress = await this.generateReceivingAddress(decoratedCallbackUrl);

    // create a new invoice
    const invoice = new Invoice({
      dueAmount,
      message,
      address: receivingAddress.address,
    });

    return invoice;
  }
}
