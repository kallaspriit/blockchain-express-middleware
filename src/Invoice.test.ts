import { Invoice, InvoiceAmountState, InvoicePaymentState } from "./";
import { processInvoiceForSnapshot } from "./Blockchain.test";

const RECEIVING_ADDRESS = "2FupTEd3PDF7HVxNrzNqQGGoWZA4rqiphq";

describe("Invoice", () => {
  it("should enable serialization and de-serialization", async () => {
    const invoice = new Invoice({
      dueAmount: 1,
      message: "Test",
      address: RECEIVING_ADDRESS,
    });

    const serialized = invoice.toJSON();
    const deSerialized = new Invoice(serialized);

    expect(processInvoiceForSnapshot(invoice)).toMatchSnapshot();
    expect(processInvoiceForSnapshot(serialized)).toMatchSnapshot();
    expect(processInvoiceForSnapshot(deSerialized)).toEqual(invoice);
  });

  it("should provide a static method for checking whether state transition is valid", async () => {
    expect(Invoice.isValidInvoiceStateTransition(InvoicePaymentState.PENDING, InvoicePaymentState.CONFIRMED)).toBe(
      true,
    );
    expect(Invoice.isValidInvoiceStateTransition(InvoicePaymentState.PENDING, InvoicePaymentState.PENDING)).toBe(true);
    expect(
      Invoice.isValidInvoiceStateTransition(InvoicePaymentState.PENDING, InvoicePaymentState.WAITING_FOR_CONFIRMATION),
    ).toBe(true);
    expect(Invoice.isValidInvoiceStateTransition(InvoicePaymentState.PENDING, InvoicePaymentState.CONFIRMED)).toBe(
      true,
    );
    expect(
      Invoice.isValidInvoiceStateTransition(
        InvoicePaymentState.WAITING_FOR_CONFIRMATION,
        InvoicePaymentState.CONFIRMED,
      ),
    ).toBe(true);
    expect(Invoice.isValidInvoiceStateTransition(InvoicePaymentState.CONFIRMED, InvoicePaymentState.PENDING)).toBe(
      false,
    );
  });

  it("should provide a static method for checking whether given state is complete", async () => {
    expect(Invoice.isCompleteState(InvoicePaymentState.PENDING)).toBe(false);
    expect(Invoice.isCompleteState(InvoicePaymentState.WAITING_FOR_CONFIRMATION)).toBe(false);
    expect(Invoice.isCompleteState(InvoicePaymentState.CONFIRMED)).toBe(true);
  });

  it("should handle registering a new transaction and update existing transactions", async () => {
    const invoice = new Invoice({
      dueAmount: 1,
      message: "Test",
      address: RECEIVING_ADDRESS,
    });

    invoice.registerTransaction({
      hash: "xxx",
      amount: 1,
      confirmations: 0,
    });

    invoice.registerTransaction({
      hash: "yyy",
      amount: 1,
      confirmations: 0,
    });

    // update first
    invoice.registerTransaction({
      hash: "xxx",
      amount: 1,
      confirmations: 1,
    });

    expect(processInvoiceForSnapshot(invoice)).toMatchSnapshot();
  });

  it("should throw if updating transaction with a different amount", async () => {
    const invoice = new Invoice({
      dueAmount: 1,
      message: "Test",
      address: RECEIVING_ADDRESS,
    });

    invoice.registerTransaction({
      hash: "xxx",
      amount: 1,
      confirmations: 0,
    });

    expect(() =>
      invoice.registerTransaction({
        hash: "xxx",
        amount: 2, // changed amount
        confirmations: 1,
      }),
    ).toThrowErrorMatchingSnapshot();
  });

  it("should handle registering a new transaction and update existing transactions", async () => {
    const invoice = new Invoice({
      dueAmount: 1,
      message: "Test",
      address: RECEIVING_ADDRESS,
    });

    invoice.registerTransaction({
      hash: "xxx",
      amount: 1,
      confirmations: 0,
    });

    invoice.registerTransaction({
      hash: "yyy",
      amount: 2,
      confirmations: 0,
    });

    // tslint:disable-next-line:no-magic-numbers
    expect(invoice.getPaidAmount()).toEqual(3);
  });

  it("should report amount state", async () => {
    const invoice = new Invoice({
      dueAmount: 5,
      message: "Test",
      address: RECEIVING_ADDRESS,
    });

    invoice.registerTransaction({
      hash: "xxx",
      amount: 2,
      confirmations: 0,
    });

    expect(invoice.getAmountState()).toEqual(InvoiceAmountState.UNDERPAID);

    invoice.registerTransaction({
      hash: "yyy",
      amount: 3,
      confirmations: 0,
    });

    expect(invoice.getAmountState()).toEqual(InvoiceAmountState.EXACT);

    invoice.registerTransaction({
      hash: "zzz",
      amount: 1,
      confirmations: 0,
    });

    expect(invoice.getAmountState()).toEqual(InvoiceAmountState.OVERPAID);
  });

  it("should accept valid payment state updates", async () => {
    const invoice = new Invoice({
      dueAmount: 5,
      message: "Test",
      address: RECEIVING_ADDRESS,
    });

    expect(invoice.getPaymentState()).toEqual(InvoicePaymentState.PENDING);

    // no change
    invoice.setPaymentState(InvoicePaymentState.PENDING);

    expect(invoice.getPaymentState()).toEqual(InvoicePaymentState.PENDING);

    // valid transition
    invoice.setPaymentState(InvoicePaymentState.WAITING_FOR_CONFIRMATION);

    expect(invoice.getPaymentState()).toEqual(InvoicePaymentState.WAITING_FOR_CONFIRMATION);

    // can't go back to pending
    expect(() => invoice.setPaymentState(InvoicePaymentState.PENDING)).toThrowErrorMatchingSnapshot();
  });

  it("should return valid signature", async () => {
    const invoice = new Invoice({
      dueAmount: 5,
      message: "Test",
      address: RECEIVING_ADDRESS,
    });

    expect(invoice.getSignature("xxx")).toMatchSnapshot();
  });

  it("should return whether current state is complete", async () => {
    const invoice = new Invoice({
      dueAmount: 5,
      message: "Test",
      address: RECEIVING_ADDRESS,
    });

    expect(invoice.isComplete()).toBe(false);

    invoice.setPaymentState(InvoicePaymentState.WAITING_FOR_CONFIRMATION);

    expect(invoice.isComplete()).toBe(false);

    invoice.setPaymentState(InvoicePaymentState.CONFIRMED);

    expect(invoice.isComplete()).toBe(true);
  });

  it("should return correct confirmation count", async () => {
    const invoice = new Invoice({
      dueAmount: 5,
      message: "Test",
      address: RECEIVING_ADDRESS,
    });

    expect(invoice.getConfirmationCount()).toBe(0);
    expect(invoice.hasSufficientConfirmations(2)).toBe(false);

    invoice.registerTransaction({
      hash: "xxx",
      amount: 1,
      confirmations: 0,
    });

    expect(invoice.getConfirmationCount()).toBe(0);
    expect(invoice.hasSufficientConfirmations()).toBe(false);

    invoice.registerTransaction({
      hash: "xxx",
      amount: 1,
      confirmations: 1,
    });

    expect(invoice.getConfirmationCount()).toBe(1);
    expect(invoice.hasSufficientConfirmations(1)).toBe(true);

    invoice.registerTransaction({
      hash: "yyy",
      amount: 1,
      confirmations: 0,
    });

    expect(invoice.getConfirmationCount()).toBe(0);
    expect(invoice.hasSufficientConfirmations()).toBe(false);

    invoice.registerTransaction({
      hash: "zzz",
      amount: 1,
      confirmations: 5,
    });

    expect(invoice.getConfirmationCount()).toBe(0);
    expect(invoice.hasSufficientConfirmations()).toBe(false);
  });
});
