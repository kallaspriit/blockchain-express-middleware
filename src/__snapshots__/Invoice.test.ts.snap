// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blockchain should accept custom logger 1`] = `
Array [
  Array [
    Object {
      "parameters": Object {
        "callback": "https://example.com?signature=4b506b7c3a37df75af9ac0a54e810ce3397fbe2255d9d820c041f9bec3a3516cfc39bce02570dfed230d81f5f37f0ba5bf886275c69d1e8b31399df45a250dd2",
        "key": "xxx",
        "xpub": "yyy",
      },
      "url": "https://api.blockchain.info/v2/receive?xpub=yyy&callback=https%3A%2F%2Fexample.com%3Fsignature%3D4b506b7c3a37df75af9ac0a54e810ce3397fbe2255d9d820c041f9bec3a3516cfc39bce02570dfed230d81f5f37f0ba5bf886275c69d1e8b31399df45a250dd2&key=xxx",
    },
    "generating receiving address",
  ],
]
`;

exports[`Blockchain should create a new invoice with receiving address 1`] = `
Object {
  "address": "2FupTEd3PDF7HVxNrzNqQGGoWZA4rqiphq",
  "createdDate": 2018-04-19T13:48:05.316Z,
  "dueAmount": 1,
  "message": "Test invoice",
  "paymentState": "PENDING",
  "stateTransitions": Array [],
  "transactions": Array [],
  "updatedDate": 2018-04-20T13:48:05.316Z,
}
`;

exports[`Blockchain should generate a new receiving address 1`] = `
Object {
  "address": "2FupTEd3PDF7HVxNrzNqQGGoWZA4rqiphq",
  "callback": "https://example.com",
  "index": 0,
}
`;

exports[`Blockchain should throw error when generating receving address and getting a non 2xx response 1`] = `[Error: Request failed with status code 400]`;

exports[`Invoice should accept valid payment state updates 1`] = `"Invalid state transition from \\"WAITING_FOR_CONFIRMATION\\" to \\"PENDING\\""`;

exports[`Invoice should enable serialization and de-serialization 1`] = `
Object {
  "address": "2FupTEd3PDF7HVxNrzNqQGGoWZA4rqiphq",
  "createdDate": 2018-04-19T13:48:05.316Z,
  "dueAmount": 1,
  "message": "Test",
  "paymentState": "PENDING",
  "stateTransitions": Array [],
  "transactions": Array [],
  "updatedDate": 2018-04-20T13:48:05.316Z,
}
`;

exports[`Invoice should enable serialization and de-serialization 2`] = `
Object {
  "address": "2FupTEd3PDF7HVxNrzNqQGGoWZA4rqiphq",
  "createdDate": 2018-04-19T13:48:05.316Z,
  "dueAmount": 1,
  "message": "Test",
  "paymentState": "PENDING",
  "stateTransitions": Array [],
  "transactions": Array [],
  "updatedDate": 2018-04-20T13:48:05.316Z,
}
`;

exports[`Invoice should handle registering a new transaction and update existing transactions 1`] = `
Object {
  "address": "2FupTEd3PDF7HVxNrzNqQGGoWZA4rqiphq",
  "createdDate": 2018-04-19T13:48:05.316Z,
  "dueAmount": 1,
  "message": "Test",
  "paymentState": "PENDING",
  "stateTransitions": Array [],
  "transactions": Array [
    Object {
      "amount": 1,
      "confirmations": 1,
      "createdDate": 2018-04-19T13:48:05.316Z,
      "hash": "xxx",
      "updatedDate": 2018-04-20T13:48:05.316Z,
    },
    Object {
      "amount": 1,
      "confirmations": 0,
      "createdDate": 2018-04-19T13:48:05.316Z,
      "hash": "yyy",
      "updatedDate": 2018-04-20T13:48:05.316Z,
    },
  ],
  "updatedDate": 2018-04-20T13:48:05.316Z,
}
`;

exports[`Invoice should return valid signature 1`] = `"c74e85fc0d8204ce05decd6ae602976f8e2cf8dcd6d07652f20584baffa557e921393fd9ad559ff6f4d9439963af80f83151535431f3946b2283dcf2bfd4316e"`;

exports[`Invoice should throw if updating transaction with a different amount 1`] = `"Invoice \\"2FupTEd3PDF7HVxNrzNqQGGoWZA4rqiphq\\" existing transaction \\"xxx\\" amount of 1 is different from the new amount of 2, this should not happen"`;
